pkg/processor/controlcommunication/controlcommunication_test.go

```diff
@@ -64,11 +65,89 @@ func (suite *ControlCommunicationTestSuite) TestSubscribeUnsubscribe() {
+	controlMessageDrain2 := make(chan *ControlMessage)
+	defer close(controlMessageDrain1)
+	defer close(controlMessageDrain2)
+
+	drainDoneConsumer, err := suite.broker.getConsumer(DrainDoneMessageKind)
+	suite.Require().NoError(err)
+
+	for _, controlMessageChan := range []chan *ControlMessage{controlMessageDrain1, controlMessageDrain2} {
+		err = suite.broker.Subscribe(DrainDoneMessageKind, controlMessageChan)
+		suite.Require().NoError(err)
+	}
+
+	defer func() {
+		err := suite.broker.Unsubscribe(StreamMessageAckKind, controlMessageDrain1)
+		suite.Require().NoError(err)
+		err = suite.broker.Unsubscribe(StreamMessageAckKind, controlMessageDrain2)
+		suite.Require().NoError(err)
+	}()
+
+	// send DrainDone message
+	suite.Require().Len(drainDoneConsumer.channels, 2)
+	err = suite.broker.SendToConsumers(&ControlMessage{Kind: DrainDoneMessageKind})
+	suite.Require().NoError(err)
+	suite.Require().Len(drainDoneConsumer.channels, 0)
+
+	message := <-controlMessageDrain1
+	suite.Require().Equal(message.Kind, DrainDoneMessageKind)
+	message = <-controlMessageDrain2
+	suite.Require().Equal(message.Kind, DrainDoneMessageKind)
+
+}
+
+func (suite *ControlCommunicationTestSuite) TestBroadcastMessage() {
+	controlMessageExplicitAck1 := make(chan *ControlMessage)
+	controlMessageExplicitAck2 := make(chan *ControlMessage)
+	defer close(controlMessageExplicitAck1)
+	defer close(controlMessageExplicitAck2)
+
+	explicitAckConsumer, err := suite.broker.getConsumer(StreamMessageAckKind)
+	suite.Require().NoError(err)
+
+	for _, controlMessageChan := range []chan *ControlMessage{controlMessageExplicitAck1, controlMessageExplicitAck2} {
+		err = suite.broker.Subscribe(StreamMessageAckKind, controlMessageChan)
+		suite.Require().NoError(err)
+	}
+	defer func() {
+		err := suite.broker.Unsubscribe(StreamMessageAckKind, controlMessageExplicitAck1)
+		suite.Require().NoError(err)
+		err = suite.broker.Unsubscribe(StreamMessageAckKind, controlMessageExplicitAck2)
+		suite.Require().NoError(err)
+	}()
+
+	suite.Require().Len(explicitAckConsumer.channels, 2)
+
+	broadcastingCtx, broadcastingDone := context.WithCancel(suite.ctx)
+	// broadcasting message (will be blocked until the message is read from all chans)
+	go func() {
+		_ = suite.broker.SendToConsumers(&ControlMessage{Kind: StreamMessageAckKind, // nolint: errcheck
+			Attributes: map[string]interface{}{"test": "broadcast"}})
+		broadcastingDone()
+	}()
+
+	for _, controlMessageChan := range []chan *ControlMessage{controlMessageExplicitAck1, controlMessageExplicitAck2} {
+		message := <-controlMessageChan
+		suite.Require().Equal(message.Kind, StreamMessageAckKind)
+		suite.Require().Equal(message.Attributes["test"], "broadcast")
+	}
+
+	select {
+	case <-broadcastingCtx.Done():
+		suite.logger.DebugWith("ExplicitAck control message was successfully broadcast")
+	case <-time.After(1 * time.Second):
+		suite.Fail("ExplicitAck control message didn't unblock consumer after message was read")
```

Comment:
if the consumer is unblocked, we might have a dangling goroutine 😱

Reply:
@TomerShor not really, because we are closing channel in defer, so we will just panic and recover.